### DISCUSSION TITLE
Is there any kernel or auxkernel solving the (\psi, f u)?

### DISCUSSION THREAD

**Discussion by hhy2022** [idaholab/moose#23177] [Q&A General] [answered]
I am trying to solve a self-defined problem. When I tranlate the strong form to the weak form, there is a term in the format like (\psi, f u), where \psi is the test function, f is a known function and u is the variable I want to solve.

I checked the lists of kernels and could not find a corresponding one to represent it. Does it mean I need to use auxkernel to represent the 'fu' and then use it as a auxvariable and then combine it to the kernel system?

I might misundertand something. If so, please correct me. 

Thank you so much for any help.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Using an auxkernel to define `f u` would make it difficult to compute the contribution to the Jacobian. 

You should use either Force or Reaction (difference is whether u is the variable being solved for in the equation) kernels here. 
There's unfornately none that I can find that uses a function for the coefficient, only either constants or material properties. 
You can convert the function to a material property using the GenericFunctionMaterial

https://mooseframework.inl.gov/source/kernels/ADMatReaction.html
https://mooseframework.inl.gov/source/kernels/MatCoupledForce.html
https://mooseframework.inl.gov/source/materials/GenericFunctionMaterial.html

Guillaume

**Reply by hhy2022** [reply]
That makes sense. Thank you so much for the information.
